explorer/interpreter/action_stack.cpp

```diff
@@ -126,24 +126,45 @@ void ActionStack::InitializeFragment(ContinuationValue::StackFragment& fragment,
   fragment.StoreReversed(std::move(reversed_todo));
 }
 
-auto ActionStack::FinishAction() -> ErrorOr<Success> {
-  std::stack<std::unique_ptr<Action>> scopes_to_destroy;
-  std::unique_ptr<Action> act = todo_.Pop();
-  switch (act->kind()) {
-    case Action::Kind::CleanUpAction:
+namespace {
+enum class FinishActionKind {
+  // Action should finish without producing a value.
+  NoValue,
+  // Action should finish by producing a value.
+  Value,
+  // Action needs custom handling. Shouldn't call FinishAction at all.
+  NoFinish,
+};
```

Comment:
I went with `NeverCalled` rather than `NeverFinishAction` because (since this is an `enum class`) `FinishActionKind` will already be part of the name so it seemed redundant to repeat. Comment changed to indicate the overall meaning; I kept the per-enumerator comments only so that I could indicate why `FinishAction` might never be called for a particular kind of `Action`.